packages/core/lib/commands/call/run.js

```diff
@@ -0,0 +1,167 @@
+    ? await sourceFromExternal(contractNameOrAddress, config)
+    : await sourceFromLocal(contractNameOrAddress, config);
+
+  ({ abi: functionEntry, tx: transaction } = await encoder.encodeTransaction(
+    functionNameOrSignature,
+    args
+  ));
+
+  if (functionEntry.stateMutability !== "view") {
+    console.log(
+      `WARNING!!! Not a view function\n  ABI stateMutability: ${functionEntry.stateMutability}`
+    );
+  }
+
+  const provider = new Encoder.ProviderAdapter(config.provider);
+  let result;
+  try {
+    result = await provider.request({
+      method: "eth_call",
+      params: [
+        {
+          from: config.from,
+          to: transaction.to,
+          data: transaction.data
+        },
+        config.blockNumber
+      ]
+    });
+  } catch (error) {
+    console.log(`Error Message: ${error.message}\nError Code: ${error.code}`);
+  }
+
+  if (result !== null && result !== undefined) {
+    const [decoding] = await decoder.decodeReturnValue(functionEntry, result);
+
+    for (const { value: result } of decoding.arguments) {
+      console.log(
+        util.inspect(new Codec.Export.ResultInspector(result), {
+          colors: true,
+          depth: null,
+          maxArrayLength: null,
+          breakLength: 79
+        })
+      );
+    }
+  }
+
+  async function sourceFromLocal(contractNameOrAddress, config) {
+    let encoder, decoder;
+    const contractNames = fs
+      .readdirSync(config.contracts_build_directory)
+      .filter(filename => filename.endsWith(".json"))
+      .map(filename => filename.slice(0, -".json".length));
+
+    const contracts = contractNames
+      .map(contractName => ({
+        [contractName]: config.resolver.require(contractName)
+      }))
+      .reduce((a, b) => ({ ...a, ...b }), {});
+
+    const isEmpty = Object.keys(contracts).length === 0;
+    if (isEmpty) {
+      throw new Error(
+        "No artifacts found! Please run `truffle compile` first to compile your contracts"
+      );
+    }
+
+    if (contractNameOrAddress.startsWith("0x")) {
+      const projectInfo = {
+        artifacts: Object.values(contracts)
+      };
+
+      ({ encoder, decoder } = await getEncoderDecoderForContractAddress(
+        contractNameOrAddress,
+        projectInfo
+      ));
+    } else {
+      const settings = {
+        provider: config.provider,
+        projectInfo: {
+          artifacts: Object.values(contracts)
+        }
+      };
+
+      const contract = contracts[contractNameOrAddress];
+      // Error handling to remind users to run truffle migrate first
+      const instance = await contract.deployed();
+      encoder = await Encoder.forContractInstance(instance, settings);
+      decoder = await Decoder.forContractInstance(instance, settings);
+    }
+    return { encoder, decoder };
+  }
+
+  async function sourceFromExternal(contractNameOrAddress, config) {
+    const { compileResult } = await fetchAndCompile(
+      contractNameOrAddress,
```

Comment:
I think you had better have some validation here on the address before calling `fetchAndCompile`!  (This comment is blocking, hitting request changes based on this.)

Reply:
I believe the validation you are mentioning on the `address` means to check if the address is 40 character hex string with "0x" prefix?

Reply:
More than that.  Addresses should be validated for a correct checksum (unless the user wants to bypass that additional layer of validation).  The usual way to validate an address is to use the `isAddress` function from web3-utils.  (I assume there's an equivalent function in ethers, too; I don't know that library as well.)  This function will check that the input both has the correct form for an address, and that its checksum... unless the address is all lowercase or all uppercase, in which case the checksum valiation is bypassed; it's understood that in these cases the user is deliberately bypassing checksum validation.  That's generally how you want to validate addresses.

Reply:
Actually, hm -- if it weren't for the necessity of accepting contract *names*, I'd say that arguably address validation should be done not with `isAddress` but rather by putting these addresses into the options (and using `allowOptions`) during encoding!  Then encoder will do the validation for you. :)  That way is probably also more extensible in the future.  And actually I guess doing it that way is perfectly viable if you introduce a separate check for contract names.

I think doing things that way does have some pitfalls, though, so you may want to talk to me if you decide to take that route.